File: VB/PatternsInAction/Patterns In Action/BusinessObjects/BusinessRules/ValidateCompare.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text

Namespace BusinessObjects.BusinessRules
	''' <summary>
	''' Compares values of two properties given a data type and operator  (>, ==, etc).
	''' </summary>
	Public Class ValidateCompare
		Inherits BusinessRule
        Private _otherPropertyName As String
        Private Property OtherPropertyName() As String
            Get
                Return _otherPropertyName
            End Get
            Set(ByVal value As String)
                _otherPropertyName = value
            End Set
        End Property
        Private _dataType As ValidationDataType
        Private Property DataType() As ValidationDataType
            Get
                Return _dataType
            End Get
            Set(ByVal value As ValidationDataType)
                _dataType = value
            End Set
        End Property
        Private _operator As ValidationOperator
        Private Property TheOperator() As ValidationOperator
            Get
                Return _operator
            End Get
            Set(ByVal value As ValidationOperator)
                _operator = value
            End Set
        End Property

        Public Sub New(ByVal propertyName As String, ByVal otherPropertyName As String, ByVal theOperator As ValidationOperator, ByVal dataType As ValidationDataType)
            MyBase.New(propertyName)

            Me.OtherPropertyName = otherPropertyName
            Me.TheOperator = theOperator
            Me.DataType = dataType

            ErrorMessage = propertyName & " must be " & theOperator.ToString() & " than " & otherPropertyName
        End Sub

		Public Sub New(ByVal propertyName As String, ByVal otherPropertyName As String, ByVal errorMessage As String, ByVal [operator] As ValidationOperator, ByVal dataType As ValidationDataType)
			Me.New(propertyName, otherPropertyName, [operator], dataType)
			Me.ErrorMessage = errorMessage
		End Sub

		Public Overrides Function Validate(ByVal businessObject As BusinessObject) As Boolean
			Try
				Dim propValue1 As String = businessObject.GetType().GetProperty(PropertyName).GetValue(businessObject, Nothing).ToString()
				Dim propValue2 As String = businessObject.GetType().GetProperty(OtherPropertyName).GetValue(businessObject, Nothing).ToString()

				Select Case DataType
					Case ValidationDataType.Integer

						Dim ival1 As Integer = Integer.Parse(propValue1)
						Dim ival2 As Integer = Integer.Parse(propValue2)

                        Select Case TheOperator
                            Case ValidationOperator.Equal
                                Return ival1 = ival2
                            Case ValidationOperator.NotEqual
                                Return ival1 <> ival2
                            Case ValidationOperator.GreaterThan
                                Return ival1 > ival2
                            Case ValidationOperator.GreaterThanEqual
                                Return ival1 >= ival2
                            Case ValidationOperator.LessThan
                                Return ival1 < ival2
                            Case ValidationOperator.LessThanEqual
                                Return ival1 <= ival2
                        End Select

					Case ValidationDataType.Double

						Dim dval1 As Double = Double.Parse(propValue1)
						Dim dval2 As Double = Double.Parse(propValue2)

                        Select Case TheOperator
                            Case ValidationOperator.Equal
                                Return dval1 = dval2
                            Case ValidationOperator.NotEqual
                                Return dval1 <> dval2
                            Case ValidationOperator.GreaterThan
                                Return dval1 > dval2
                            Case ValidationOperator.GreaterThanEqual
                                Return dval1 >= dval2
                            Case ValidationOperator.LessThan
                                Return dval1 < dval2
                            Case ValidationOperator.LessThanEqual
                                Return dval1 <= dval2
                        End Select

					Case ValidationDataType.Decimal

						Dim cval1 As Decimal = Decimal.Parse(propValue1)
						Dim cval2 As Decimal = Decimal.Parse(propValue2)

                        Select Case TheOperator
                            Case ValidationOperator.Equal
                                Return cval1 = cval2
                            Case ValidationOperator.NotEqual
                                Return cval1 <> cval2
                            Case ValidationOperator.GreaterThan
                                Return cval1 > cval2
                            Case ValidationOperator.GreaterThanEqual
                                Return cval1 >= cval2
                            Case ValidationOperator.LessThan
                                Return cval1 < cval2
                            Case ValidationOperator.LessThanEqual
                                Return cval1 <= cval2
                        End Select

					Case ValidationDataType.Date

						Dim tval1 As DateTime = DateTime.Parse(propValue1)
						Dim tval2 As DateTime = DateTime.Parse(propValue2)

                        Select Case TheOperator
                            Case ValidationOperator.Equal
                                Return tval1 = tval2
                            Case ValidationOperator.NotEqual
                                Return tval1 <> tval2
                            Case ValidationOperator.GreaterThan
                                Return tval1 > tval2
                            Case ValidationOperator.GreaterThanEqual
                                Return tval1 >= tval2
                            Case ValidationOperator.LessThan
                                Return tval1 < tval2
                            Case ValidationOperator.LessThanEqual
                                Return tval1 <= tval2
                        End Select

					Case ValidationDataType.String

						Dim result As Integer = String.Compare(propValue1, propValue2, StringComparison.CurrentCulture)

                        Select Case TheOperator
                            Case ValidationOperator.Equal
                                Return result = 0
                            Case ValidationOperator.NotEqual
                                Return result <> 0
                            Case ValidationOperator.GreaterThan
                                Return result > 0
                            Case ValidationOperator.GreaterThanEqual
                                Return result >= 0
                            Case ValidationOperator.LessThan
                                Return result < 0
                            Case ValidationOperator.LessThanEqual
                                Return result <= 0
                        End Select

				End Select
				Return False
			Catch
				Return False
			End Try
		End Function
	End Class
End Namespace

File: CSharp/PatternsInAction/Patterns In Action/BusinessObjects/BusinessRules/ValidateCompare.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BusinessObjects.BusinessRules
{
    /// <summary>
    /// Compares values of two properties given a data type and operator  (>, ==, etc).
    /// </summary>
    public class ValidateCompare : BusinessRule
    {
        private string OtherPropertyName { get; set; }
        private ValidationDataType DataType { get; set; }
        private ValidationOperator Operator { get; set; }

        public ValidateCompare(string propertyName, string otherPropertyName, 
            ValidationOperator @operator, ValidationDataType dataType )
            : base(propertyName)
        {
            
            OtherPropertyName = otherPropertyName;
            Operator = @operator;
            DataType = dataType;

            ErrorMessage = propertyName + " must be " + Operator.ToString() + " than " + otherPropertyName;
        }

        public ValidateCompare(string propertyName, string otherPropertyName, string errorMessage,
            ValidationOperator @operator, ValidationDataType dataType )
            : this(propertyName, otherPropertyName, @operator, dataType)
        {
            ErrorMessage = errorMessage;
        }

        public override bool Validate(BusinessObject businessObject)
        {
            try
            {
                string propValue1 = businessObject.GetType().GetProperty(PropertyName).GetValue(businessObject, null).ToString();
                string propValue2 = businessObject.GetType().GetProperty(OtherPropertyName).GetValue(businessObject, null).ToString();
                
                switch(DataType)
                {
                    case ValidationDataType.Integer:

                        int ival1 = int.Parse(propValue1);
                        int ival2 = int.Parse(propValue2);

                        switch(Operator)
                        {
                            case ValidationOperator.Equal: return ival1 == ival2; 
                            case ValidationOperator.NotEqual: return ival1 != ival2; 
                            case ValidationOperator.GreaterThan: return ival1 > ival2; 
                            case ValidationOperator.GreaterThanEqual: return ival1 >= ival2; 
                            case ValidationOperator.LessThan: return ival1 < ival2; 
                            case ValidationOperator.LessThanEqual: return ival1 <= ival2; 
                        }
                        break;
                        
                    case ValidationDataType.Double:

                        double dval1 = double.Parse(propValue1);
                        double dval2 = double.Parse(propValue2);

                        switch(Operator)
                        {
                            case ValidationOperator.Equal: return dval1 == dval2; 
                            case ValidationOperator.NotEqual: return dval1 != dval2; 
                            case ValidationOperator.GreaterThan: return dval1 > dval2; 
                            case ValidationOperator.GreaterThanEqual: return dval1 >= dval2; 
                            case ValidationOperator.LessThan: return dval1 < dval2; 
                            case ValidationOperator.LessThanEqual: return dval1 <= dval2; 
                        }
                        break;
                        
                    case ValidationDataType.Decimal:

                        decimal cval1 = decimal.Parse(propValue1);
                        decimal cval2 = decimal.Parse(propValue2);

                        switch(Operator)
                        {
                            case ValidationOperator.Equal: return cval1 == cval2; 
                            case ValidationOperator.NotEqual: return cval1 != cval2; 
                            case ValidationOperator.GreaterThan: return cval1 > cval2; 
                            case ValidationOperator.GreaterThanEqual: return cval1 >= cval2; 
                            case ValidationOperator.LessThan: return cval1 < cval2; 
                            case ValidationOperator.LessThanEqual: return cval1 <= cval2; 
                        }
                        break;

                    case ValidationDataType.Date:

                        DateTime tval1 = DateTime.Parse(propValue1);
                        DateTime tval2 = DateTime.Parse(propValue2);

                        switch(Operator)
                        {
                            case ValidationOperator.Equal: return tval1 == tval2; 
                            case ValidationOperator.NotEqual: return tval1 != tval2; 
                            case ValidationOperator.GreaterThan: return tval1 > tval2; 
                            case ValidationOperator.GreaterThanEqual: return tval1 >= tval2; 
                            case ValidationOperator.LessThan: return tval1 < tval2; 
                            case ValidationOperator.LessThanEqual: return tval1 <= tval2; 
                        }
                        break;

                    case ValidationDataType.String:

                        int result = string.Compare(propValue1, propValue2, StringComparison.CurrentCulture);

                        switch(Operator)
                        {
                            case ValidationOperator.Equal: return result == 0; 
                            case ValidationOperator.NotEqual: return result != 0; 
                            case ValidationOperator.GreaterThan: return result > 0; 
                            case ValidationOperator.GreaterThanEqual: return result >= 0; 
                            case ValidationOperator.LessThan: return result < 0; 
                            case ValidationOperator.LessThanEqual: return result <= 0; 
                        }
                        break;

                }
                return false;
            }
            catch{ return false; }
        }
    }
}
